### PR TITLE
Lazy Formatter

### PR DESCRIPTION
**Pull request by zrthxn**

### CODE REVIEW

File: src/lazy/lazy.cpp [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+int main() {
+  auto val = lazy([&]{
+    return UINT16_MAX;
+  });
+  val();
+
+  // auto const cval = concurrent_lazy([&]{
+  //   return UINT32_MAX;
+  // });
+  // cval();
```
Comment (a20012251) [inline src/lazy/lazy.cpp:16]: concurrent lazy should be in a different file and PR

**Review by a20012251 — COMMENTED** [review]
looking good!

File: src/lazy/lazy.cpp (hunk 1 above)
Comment (zrthxn) [reply src/lazy/lazy.cpp:16]: Yes I was just trying it out, I think pretty much the same formatter will work. I couldn't get this to compile though, it gives a difficult to understand error...

Comment (a20012251) [reply src/lazy/lazy.cpp:16]: Let me know if you need help compiling it

File: src/lazy/lazy.cpp [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+  uint runs = 0;
+  auto val = lazy([&]{
+    // Expensive value calculation
+    uint16_t _i = 0;
+    ++runs;
+    while (_i < UINT16_MAX) {
+      ++_i;
+    }
+    return (uint32_t)(_i + runs);
+  });
```
Comment (a20012251) [inline src/lazy/lazy.cpp:16]: let's better do something simpler like this
```
int i = 10;
auto val = lazy([&]{
  return i++;
});
// at this point val will return 11 only if the callback is invoked twice

  // These should be equal if the calculation is run once
  std::cout << "First call: " << val() << std::endl;
  std::cout << "Reused call: " << val() << std::endl;
```

File: src/lazy/lazy_formatter.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+		return value
+	
+	def update(self):
+		self.has_value = True if (self.valobj
+			.GetChildMemberWithName('value_')
+			.GetChildMemberWithName('storage_')
+			.GetChildMemberWithName('hasValue')
+			.GetValueAsUnsigned()
+		) == 1 else False
```
Comment (a20012251) [inline src/lazy/lazy_formatter.py:35]: ) != 0 else False

this is safer because you can't always guarantee that a `true` value will be 1.

**Review by a20012251 — CHANGES_REQUESTED** [review]
minor changes needed

File: src/lazy/lazy_formatter.py [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+		debugger.HandleCommand(
+			'type synthetic add '
+			+ f'-x "{typeName}" '
+			+ f'--python-class {moduleName}.LazyFormatter'
+		)
+
+		debugger.HandleCommand(
+			'type summary add --expand ' 
+			+ f'-x "{typeName}" ' 
+			+ f'--summary-string "Initialized=${{var.value_.storage_.hasValue}}"'
```
Comment (a20012251) [inline src/lazy/lazy_formatter.py:56]: Let's better use the word 'Calculated' instead of Initialized.

**Review by a20012251 — CHANGES_REQUESTED** [review]
minor changes needed

File: src/lazy/lazy_formatter.py (hunk 4 above)
Comment (zrthxn) [reply src/lazy/lazy_formatter.py:56]: Alright but "calculated" sounds like a mathematical value being computed. I think some other word which implies that a value is has been fetched or set would be better.

**Comment by zrthxn** [comment]
```
Traceback (most recent call last):
  File "/home/alisamar/Desktop/lldb_folly_formatter/src/lazy/lazy_formatter.py", line 45, in LazySummary
    computed = bool(lldb.value(valobj).value_.storage_.hasValue)
  File "/usr/lib64/python3.9/site-packages/lldb/__init__.py", line 14445, in __getattr__
    raise AttributeError("Attribute '%s' is not defined" % name)
AttributeError: Attribute 'value_' is not defined
```

**Comment by zrthxn** [comment]
Tests pass!

**Review by a20012251 — APPROVED** [review]
amazing!

**Pull request merged**